lib/shaders/pathtracer_compute.wgsl

```diff
@@ -85,190 +84,139 @@ fn traceRay(ray: Ray, seed: u32) -> vec3f {
-        var v2: vec3f = vec3f(vertices[i+6], vertices[i+7], vertices[i+8]);
-
-        var t: f32 = hitTriangle(ray, v0, v1, v2);
-        if (t > 0.0 && (t < hitInfo.t || hitInfo.t < 0.0)) {
-            newHitInfo.t = t;
-
-            // Calculate the normal using the cross product of the edge vectors
-            var edge1: vec3<f32> = v1 - v0;
-            var edge2: vec3<f32> = v2 - v0;
-            newHitInfo.normal = normalize(cross(edge1, edge2));
-
-            newHitInfo.color = vec3f(1.0, 0.1, 0.1);
-            newHitInfo.emissiveColor = vec3f(0.0, 0.0, 0.0);
-        }
+fn intersectXZPlane(
+    y: f32,
+    material: ptr<function, Material>,
+    ray: Ray,
+    bestHit: ptr<function, HitInfo>) {
+
+    var xzPlane: XZPlane;
+    xzPlane.normal = vec3f(0.0, 1.0, 0.0);
+    xzPlane.distance = -y;
+
+    let t: f32 = -(dot(ray.origin, xzPlane.normal) + xzPlane.distance) / dot(ray.direction, xzPlane.normal);
+    if (t > 0.0 && t < (*bestHit).t) {
+        (*bestHit).t = t;
+        (*bestHit).normal = vec3f(0.0, 1.0, 0.0);
+        (*bestHit).material = *(material); // TODO: reference directly or use index
     }
-
-    return newHitInfo;
 }
 
-fn hitFloor(ray: Ray, floor: Floor) -> f32 {
-    var t: f32 = -(dot(ray.origin, floor.normal) + floor.distance) / dot(ray.direction, floor.normal);
-    if (t > 0.0) {
-        return t;
-    }
-    return -1.0;
-}
+fn intersectSphere(
+    sphere: ptr<function, Sphere>,
+    material: ptr<function, Material>,
+    ray: Ray,
+    bestHit: ptr<function, HitInfo>) {
 
-fn hitSphere(ray: Ray, sphere: Sphere) -> f32 {
-    
     let a: f32 = dot(ray.direction, ray.direction);
-    let b: f32 = 2.0 * dot(ray.direction, ray.origin - sphere.center);
-    let c: f32 = dot(ray.origin - sphere.center, ray.origin - sphere.center) - sphere.radius * sphere.radius;
+    let b: f32 = 2.0 * dot(ray.direction, ray.origin - (*sphere).center);
+    let c: f32 = dot(ray.origin - (*sphere).center, ray.origin - (*sphere).center) - (*sphere).radius * (*sphere).radius;
     let discriminant: f32 = b * b - 4.0 * a * c;
 
     if(discriminant < 0.0) {
-        return -1.0;
+        return;
     }
 
     let t0: f32 = (-b - sqrt(discriminant)) / (2.0 * a);
     let t1: f32 = (-b + sqrt(discriminant)) / (2.0 * a);
 
-    return min(t0, t1);
+    let t: f32 = min(t0, t1);
+
+    if (t > 0.0 && t < (*bestHit).t) {
+        (*bestHit).t = t;
+        (*bestHit).normal = normalize(rayAt(ray, t) - (*sphere).center);
+        (*bestHit).material = (*material); // TODO: reference directly or use index
+    }
 }
 
-fn hitTriangle(ray: Ray, v0: vec3<f32>, v1: vec3<f32>, v2: vec3<f32>) -> f32 {
-    const EPSILON: f32 = 0.0000001;
-    
-    var vertex0: vec3<f32> = v0;
-    var vertex1: vec3<f32> = v1;
-    var vertex2: vec3<f32> = v2;
+fn intersectTriangles(ray: Ray, bestHit: ptr<function, HitInfo>) {
+    for(var i: u32 = 0; i < 8712; i+=9) { // TODO: pass triangle count as uniform 
```

Comment:
We can use arrayLength()

Reply:
@umuterc, I did not know that such a method exists. I will use that.